File: src/main/java/com/d20charactersheet/framework/dac/dao/sql/rowmapper/KnownSpellRowMapper.java
package com.d20charactersheet.framework.dac.dao.sql.rowmapper;

import java.sql.SQLException;
import java.util.List;

import org.jetbrains.annotations.NotNull;

import com.d20charactersheet.framework.boc.model.KnownSpell;
import com.d20charactersheet.framework.boc.model.Spell;
import com.d20charactersheet.framework.boc.model.Spelllist;
import com.d20charactersheet.framework.dac.dao.sql.DataRow;
import com.d20charactersheet.framework.dac.dao.sql.RowMapper;

/**
 * Maps raw data to KnownSpell class.
 */
public class KnownSpellRowMapper implements RowMapper {

    private final List<Spell> allSpells;
    private final List<Spelllist> allSpelllists;

    /**
     * Instantiates KnownSpellRowMapper with all spell lists and spells of the game system.
     *
     * @param allSpelllists
     *     All spell lists of the game system.
     * @param allSpells
     *     All spells of the game system.
     */
    public KnownSpellRowMapper(final List<Spelllist> allSpelllists, final List<Spell> allSpells) {
        this.allSpelllists = allSpelllists;
        this.allSpells = allSpells;
    }

    @NotNull
    @Override
    public Object mapRow(@NotNull DataRow dataRow) throws SQLException {
        final KnownSpell knownSpell = new KnownSpell();
        knownSpell.setId(dataRow.getInt(0));
        knownSpell.setSpelllist(getSpelllist(dataRow.getInt(1)));
        knownSpell.setSpell(getSpell(dataRow.getInt(2)));
        return knownSpell;
    }

    private Spelllist getSpelllist(final int spelllistId) {
        for (final Spelllist spelllist : allSpelllists) {
            if (spelllist.getId() == spelllistId) {
                return spelllist;
            }
        }
        return null;
    }

    private Spell getSpell(final int spellId) {
        for (final Spell spell : allSpells) {
            if (spell.getId() == spellId) {
                return spell;
            }
        }
        throw new IllegalArgumentException("Can't find spell with id: " + spellId);
    }

}

File: src/main/java/com/d20charactersheet/framework/dac/dao/sql/rowmapper/KnownSpellRowMapper.kt
package com.d20charactersheet.framework.dac.dao.sql.rowmapper

import com.d20charactersheet.framework.boc.model.KnownSpell
import com.d20charactersheet.framework.boc.model.Spell
import com.d20charactersheet.framework.boc.model.Spelllist
import com.d20charactersheet.framework.dac.dao.sql.DataRow
import com.d20charactersheet.framework.dac.dao.sql.RowMapper
import java.sql.SQLException

/**
 * Instantiates KnownSpellRowMapper with all spell lists and spells of the game system.
 *
 * @param allSpelllists
 * All spell lists of the game system.
 * @param allSpells
 * All spells of the game system.
 */
class KnownSpellRowMapper(private val allSpelllists: List<Spelllist>, private val allSpells: List<Spell>) : RowMapper {

    /**
     * Maps raw data to KnownSpell class.
     */
    @Throws(SQLException::class)
    override fun mapRow(dataRow: DataRow): Any {
        val knownSpell = KnownSpell()
        knownSpell.id = dataRow.getInt(0)
        knownSpell.spelllist = getSpelllist(dataRow.getInt(1))
        knownSpell.spell = getSpell(dataRow.getInt(2))
        return knownSpell
    }

    private fun getSpelllist(spelllistId: Int): Spelllist? {
        for (spelllist in allSpelllists) {
            if (spelllist.id == spelllistId) {
                return spelllist
            }
        }
        return null
    }

    private fun getSpell(spellId: Int): Spell {
        for (spell in allSpells) {
            if (spell.id == spellId) {
                return spell
            }
        }
        throw IllegalArgumentException("Can't find spell with id: $spellId")
    }
}